### PR TITLE
Gateway: Adds header optimization

### PR DESCRIPTION
**Pull request by j82w**
# Pull Request Template

## Description

This adds a optimized header for the gateway store model. This model wraps the existing header objects. This avoids iterating and doing string join on all the operations.

Top10AVGRps for gateway mode
--

Master: 11868.333333
Optimized: 12555.411765

Profiler shows it consumes %16 of CPU from code in main branch. After this PR the profiler no longer shows the conversion.
MicrosoftTeams-image (1)

## Type of change

Please delete options that are not relevant.

- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] This change requires a documentation update

## Closing issues

To automatically close an issue: closes #IssueNumber

### CODE REVIEW

**Comment by kirankumarkolli** [comment]
What's the RPS impact?

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/FeedToken/ChangeFeedIteratorCoreTests.cs [hunk 1]
```diff
@@ -148,7 +148,6 @@ await feedIterator.ReadNextAsync(this.cancellationToken))
         /// </summary>
         /// <returns></returns>
         [TestMethod]
-        [Timeout(30000)]
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/FeedToken/ChangeFeedIteratorCoreTests.cs:151]: Is it related? 

If not can it go as separate PR?

Comment (j82w) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/FeedToken/ChangeFeedIteratorCoreTests.cs:151]: I will revert in next iteration.

**Comment by j82w** [comment]
> What's the RPS impact?

Still running RPS for direct mode to see if there is an impact. Gateway mode RPS is described in the PR description with roughly %5.6 gain.

File: Microsoft.Azure.Cosmos/src/Headers/HttpResponseHeadersWrapper.cs [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+            if (responseHeaders.TryGetValues(HttpConstants.HttpHeaders.OwnerFullName, out IEnumerable<string> values))
+            {
+                responseHeaders.Remove(HttpConstants.HttpHeaders.OwnerFullName);
+                foreach (string val in values)
+                {
+                    responseHeaders.Add(HttpConstants.HttpHeaders.OwnerFullName, Uri.UnescapeDataString(val));
+                }
+            }
+
+            this.httpResponseHeaders = responseHeaders;
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Headers/HttpResponseHeadersWrapper.cs:42]: Since now we are not copying the headers, if the HttpResponseMessage is disposed, would that dispose the Headers?

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Headers/HttpResponseHeadersWrapper.cs:42]: No, the headers do not implement IDisposable.

File: Microsoft.Azure.Cosmos/src/Headers/HttpResponseHeadersWrapper.cs [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+        /// This optimizes to combine the 2 headers without iterating overall of them to duplicate it into a new
+        /// header object.
+        /// </summary>
+        /// <param name="responseHeaders"></param>
+        /// <param name="httpContentHeaders"></param>
+        public HttpResponseHeadersWrapper(
+            HttpResponseHeaders responseHeaders,
+            HttpContentHeaders httpContentHeaders)
+        {
+            if (responseHeaders.TryGetValues(HttpConstants.HttpHeaders.OwnerFullName, out IEnumerable<string> values))
```
[the remark this answers is not in the document]
Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Headers/HttpResponseHeadersWrapper.cs:33]: The content headers should never have the OwnerFullName. I think who ever originally implemented it did not understand the difference between the two headers.

File: Microsoft.Azure.Cosmos/src/Headers/HttpResponseHeadersWrapper.cs (hunk 2 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Headers/HttpResponseHeadersWrapper.cs:42]: But if the HttpResponseMessage is disposed and the Cosmos.ResponseMessage is maintained, does the reference to the original headers make the HttpResponseMessage not to be disposed until the Cosmos.ResponseMessage is disposed?

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Headers/HttpResponseHeadersWrapper.cs:42]: No, the original `HttpResponseMessage` is always disposed of when it is converted to a `DocumentServiceResponse`. The `HttpResponseHeaders` does not have a reference to the original `HttpResponseMessage` so it won't stop garbage cleanup either. 
https://github.com/Azure/azure-cosmos-dotnet-v3/blob/33c0e9a1a548b7fcef0a3607e4c206e8220e2169/Microsoft.Azure.Cosmos/src/GatewayStoreClient.cs#L91

File: Microsoft.Azure.Cosmos/src/Headers/HttpResponseHeadersWrapper.cs [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+                }
+            }
+
+            return headers;
+        }
+
+        public override int Count()
+        {
+            int count = 0;
+            if (this.dictionaryNameValueCollection.IsValueCreated)
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Headers/HttpResponseHeadersWrapper.cs:111]: Nice:smile:

File: Microsoft.Azure.Cosmos/src/Headers/HttpResponseHeadersWrapper.cs [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+            if (!this.httpResponseHeaders.TryGetValues(key, out IEnumerable<string> result))
+            {
+                if (this.httpContentHeaders == null || !this.httpContentHeaders.TryGetValues(key, out result))
+                {
+                    if (this.dictionaryNameValueCollection.IsValueCreated)
+                    {
+                        return this.dictionaryNameValueCollection.Value.Get(key);
+                    }
+                }
+            }
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Headers/HttpResponseHeadersWrapper.cs:135]: assuming an internal functional wrapper is okey, will code readability be better like below

IEnum<string> GetHeader(string key) {
   if (this....TryGetValue(..., out ... value)) {
      return value;
   }
}

**Review by ealsur — APPROVED** [review]
